plugin-server/src/cdp/templates/serve-native-destination.ts

```diff
@@ -0,0 +1,144 @@
+        mockFn.mockResolvedValue = (value: any) => {
+            mockFn.mockReturnValue = () => Promise.resolve(value)
+            return mockFn
+        }
+        mockFn.mockReturnValue = (value: any) => {
+            mockFn.mockReturnValue = () => value
+            return mockFn
+        }
+        return mockFn
+    },
+    spyOn: (obj: any, method: string) => {
+        const original = obj[method]
+        const mockFn: any = (global as any).jest.fn()
+        obj[method] = mockFn
+        mockFn.mockRestore = () => {
+            obj[method] = original
+        }
+        return mockFn
+    },
+    useRealTimers: () => {},
+}
+
+import cors from 'cors'
+import express from 'express'
+import inquirer from 'inquirer'
+
+import { NATIVE_HOG_FUNCTIONS } from './index'
+import { DestinationTester } from './test/test-helpers'
+
+const PORT = 4321
+
+async function main() {
+    const { hogFunction: hogFunctionId } = await inquirer.prompt([
+        {
+            type: 'list',
+            name: 'hogFunction',
+            message: 'Select a hog function:',
+            choices: NATIVE_HOG_FUNCTIONS.filter((hogFunction) => hogFunction.id !== 'native-dev-center').map(
+                (hogFunction) => ({
+                    name: hogFunction.name,
+                    value: hogFunction.id,
+                })
+            ),
+        },
+    ])
+
+    const hogFunction = NATIVE_HOG_FUNCTIONS.find((hogFunction) => hogFunction.id === hogFunctionId)
+
+    if (!hogFunction) {
+        throw new Error('You must select a destination. Exiting...')
+    }
+    const { mapping: mappingName } = await inquirer.prompt([
+        {
+            type: 'list',
+            name: 'mapping',
+            message: 'Select a mapping:',
+            choices: hogFunction.mapping_templates.map((mapping) => ({
+                name: mapping.name,
+                value: mapping.name,
+            })),
+        },
+    ])
+
+    const mapping = hogFunction.mapping_templates.find((mapping) => mapping.name === mappingName)
+
+    if (!mapping) {
+        throw new Error('You must select a mapping. Exiting...')
+    }
+
+    console.log(`You selected: ${hogFunctionId} / ${mapping.name}`)
+
+    const app = express()
+
+    // Enable CORS for localhost
+    app.use(
+        cors({
+            origin: [
+                'http://localhost:8000',
+                'http://localhost:8010',
+                'https://us.posthog.com',
+                'https://eu.posthog.com',
+            ],
+        })
+    )
+
+    app.use(express.json())
+
+    app.get('/local-hog-function', (req, res) => {
+        res.status(200).json({
+            ...hogFunction,
+            mapping_templates: hogFunction.mapping_templates.filter((m) => m.name === mapping.name),
+        })
+    })
+
+    app.post('/local-hog-function/invoke', async (req, res) => {
+        const { globals, configuration } = req.body
+        const tester = new DestinationTester(hogFunction)
```

Comment:
if i get it, this means the user will end up needing to run the whole backend stack successfully in order to test any of this. Something tells me this is going to end up being so hard anyways that few people would do it... There's also going to be a tonne of edge cases here such as secret variables or integrations (like slack) that definitely won't work. Again I like the idea but I feel like the usefulness of this will be _pretty_ limited.

The risk here is that this tool makes it more confusing to build a destination rather than less... Again, happy to move forward here but there is a lot of missing parts here that might not have an easy fix, especially going in the direction you are so far. Something to think about...

Reply:
It just is super super limited though. Integrations won't work at all, you still need all of the local modules built and installed (pretty sure this wont start up otherwise). Also has some weird security considerations of the server talking to localhost (i'm not even sure that it works at all on production once we have the CSP in place).

We can try it but my gut tells me this is going to be a bit of a mess to manage...

Reply:
Also there is no hog compliation so none of that will work... To make that work you end up almost needing the whole stack anyways.... Again just feels all a little... messy...